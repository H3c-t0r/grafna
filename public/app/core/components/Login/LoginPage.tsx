import React, { FC } from 'react';
import config from 'app/core/config';
import { UserSignup } from './UserSignup';
import { LoginServiceButtons } from './LoginServiceButtons';
import LoginCtrl from './LoginCtrl';
import { LoginForm } from './LoginForm';
import { ChangePassword } from './ChangePassword';
import { CSSTransition } from 'react-transition-group';

const isOauthEnabled = () => Object.keys(config.oauth).length > 0;

export const LoginPage: FC = () => {
  return (
    <div className="login container">
      <div className="login-content">
        <div className="login-branding">
          <img className="logo-icon" src="public/img/grafana_icon.svg" alt="Grafana" />
          <div className="logo-wordmark" />
        </div>
        <LoginCtrl>
          {({ login, isLoggingIn, changePassword, toGrafana, isChangingPassword }) => (
            <div className="login-out-box">
              <div className={`login-inner-box ${isChangingPassword ? 'hidden' : ''}`} id="login-view">
                <LoginForm
                  displayLoginFields={!config.disableLoginForm}
                  displayForgotPassword={!(config.ldapEnabled || config.authProxyEnabled)}
                  onSubmit={login}
                  loginHint={config.loginHint}
                  passwordHint={config.passwordHint}
                  isLoggingIn={isLoggingIn}
                />

                {isOauthEnabled() ? (
                  <>
                    <div className="text-center login-divider">
                      <div>
                        <div className="login-divider-line" />
                      </div>
                      <div>
                        <span className="login-divider-text">{config.disableLoginForm ? null : <span>or</span>}</span>
                      </div>
                      <div>
                        <div className="login-divider-line" />
                      </div>
                    </div>
                    <div className="clearfix" />

                    <LoginServiceButtons />
                  </>
                ) : null}

                <UserSignup />
              </div>
              <CSSTransition appear={true} in={isChangingPassword} timeout={250} classNames="login-inner-box">
                <ChangePassword
                  onSubmit={changePassword}
                  onSkip={toGrafana}
                  focus={isChangingPassword}
                  className={isChangingPassword ? '' : 'hidden'}
                />
              </CSSTransition>
            </div>
          )}
        </LoginCtrl>

        <div className="clearfix" />
      </div>
    </div>
  );
};
