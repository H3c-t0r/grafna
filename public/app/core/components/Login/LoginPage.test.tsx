import React from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { LoginPage } from './LoginPage';

const postMock = jest.fn();
jest.mock('@grafana/runtime', () => ({
  getBackendSrv: () => ({
    post: postMock,
  }),
}));

jest.mock('app/core/config', () => {
  return {
    loginError: false,
    buildInfo: {
      version: 'v1.0',
      commit: '1',
      env: 'production',
      edition: 'Open Source',
      isEnterprise: false,
    },
    licenseInfo: {
      stateInfo: '',
      licenseUrl: '',
    },
    appSubUrl: '',
    getConfig: () => ({
      appSubUrl: '',
      verifyEmailEnabled: false,
    }),
  };
});

describe('Login Page', () => {
  it('renders correctly', () => {
    const { container } = render(<LoginPage />);
    expect(container).toMatchSnapshot();
  });
  it('should pass validation checks for username field', async () => {
    render(<LoginPage />);
    // when
    await act(async () => {
      fireEvent.click(screen.getByLabelText('Login button'));
    });
    // then
    expect(screen.queryByText('Email or username is required')).toBeInTheDocument();
    // when
    await act(async () => {
      userEvent.type(screen.getByLabelText('Username input field'), 'admin');
    });
    // then
    expect(screen.queryByText('Email or username is required')).not.toBeInTheDocument();
  });
  it('should pass validation checks for password field', async () => {
    render(<LoginPage />);
    // when
    await act(async () => {
      fireEvent.click(screen.getByLabelText('Login button'));
    });
    // then
    expect(screen.queryByText('Password is required')).toBeInTheDocument();
    // when
    await act(async () => {
      userEvent.type(screen.getByLabelText('Password input field'), 'admin');
    });
    // then
    expect(screen.queryByText('Password is required')).not.toBeInTheDocument();
  });
  it('should navigate to default url if credentials is valid', async () => {
    Object.defineProperty(window, 'location', {
      value: {
        assign: jest.fn(),
      },
    });
    postMock.mockResolvedValueOnce({ message: 'Logged in' });
    render(<LoginPage />);
    // when
    await act(async () => {
      userEvent.type(screen.getByLabelText('Username input field'), 'admin');
      userEvent.type(screen.getByLabelText('Password input field'), 'test');
      fireEvent.click(screen.getByLabelText('Login button'));
    });
    // then
    expect(postMock).toHaveBeenCalledWith('/login', { password: 'test', user: 'admin' });
    expect(window.location.assign).toHaveBeenCalledWith('/');
  });
});
