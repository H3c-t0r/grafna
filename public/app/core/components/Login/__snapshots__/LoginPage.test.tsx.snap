// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page renders correctly 1`] = `
<div>
  <div
    class="css-1cgt561"
  >
    <div
      class="login-content-box css-odanoj"
    >
      <div
        class="css-1yyz87g"
      >
        <img
          alt="Grafana"
          class="css-1px3c36"
          src="public/img/grafana_icon.svg"
        />
        <div
          class="css-1w4npsm"
        >
          <h1
            class="css-vxdt2o"
          >
            Welcome to Grafana
          </h1>
        </div>
      </div>
      <div
        class="css-9h8xxw"
      >
        <div
          class="css-iim5di"
        >
          <div
            class="css-lcb2lo"
          >
            <form
              class="css-xs0vux"
            >
              <div
                class="css-8e5b3"
              >
                <div
                  class="css-jt4xma-Label"
                >
                  <label>
                    <div
                      class="css-xhqy0o"
                    >
                      Email or username
                    </div>
                  </label>
                </div>
                <div>
                  <div
                    class="css-xcstkt-input-wrapper"
                  >
                    <div
                      class="css-1w5c5dq-input-inputWrapper"
                    >
                      <input
                        aria-label="Username input field"
                        autocapitalize="none"
                        class="css-mtvhwr-input-input"
                        name="user"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="css-8e5b3"
              >
                <div
                  class="css-jt4xma-Label"
                >
                  <label>
                    <div
                      class="css-xhqy0o"
                    >
                      Password
                    </div>
                  </label>
                </div>
                <div>
                  <div
                    class="css-xcstkt-input-wrapper"
                  >
                    <div
                      class="css-1w5c5dq-input-inputWrapper"
                    >
                      <input
                        aria-label="Password input field"
                        class="css-mtvhwr-input-input"
                        name="password"
                        type="password"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <button
                aria-label="Login button"
                class="css-w9m50q-button"
              >
                <span
                  class="css-1mhnkuh"
                >
                  Log in
                </span>
              </button>
              <div
                class="css-4bnqyk-horizontal-group"
                style="width: 100%; height: 100%;"
              >
                <div
                  class="css-3h7b25-layoutChildrenWrapper"
                >
                  <a
                    class="css-19vj0lr-button"
                    href="/user/password/send-reset-email"
                    tabindex="0"
                  >
                    <span
                      class="css-1mhnkuh"
                    >
                      Forgot your password?
                    </span>
                  </a>
                </div>
              </div>
            </form>
          </div>
          <div
            class="css-66nrdr-vertical-group"
            style="width: 100%; height: 100%;"
          >
            <div
              class="css-1708ggn-layoutChildrenWrapper"
            >
              <div
                class="css-1v3caum"
              >
                New to Grafana?
              </div>
            </div>
            <div
              class="css-1708ggn-layoutChildrenWrapper"
            >
              <a
                class="css-1ulskws-button"
                href="/signup"
                tabindex="0"
              >
                <span
                  class="css-1mhnkuh"
                >
                  Sign up
                </span>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="footer"
    >
      <div
        class="text-center"
      >
        <ul>
          <li>
            <a
              href="https://grafana.com/docs/grafana/latest/?utm_source=grafana_footer"
              rel="noopener"
              target="_blank"
            >
              <div
                class="css-1vzus6i-Icon"
              />
               
              Documentation
            </a>
          </li>
          <li>
            <a
              href="https://grafana.com/products/enterprise/?utm_source=grafana_footer"
              rel="noopener"
              target="_blank"
            >
              <div
                class="css-1vzus6i-Icon"
              />
               
              Support
            </a>
          </li>
          <li>
            <a
              href="https://community.grafana.com/?utm_source=grafana_footer"
              rel="noopener"
              target="_blank"
            >
              <div
                class="css-1vzus6i-Icon"
              />
               
              Community
            </a>
          </li>
          <li>
            <a
              href=""
              rel="noopener"
            >
               
              Open Source
            </a>
          </li>
          <li>
            <a
              rel="noopener"
            >
               
              vv1.0 (1)
            </a>
          </li>
        </ul>
      </div>
    </footer>
  </div>
</div>
`;
